proto/api.go

```diff
@@ -51,41 +52,187 @@ const (
 	isWrite                // write cmds go through raft and must be proposed on leader
 	isTxnWrite             // txn write cmds start heartbeat and are marked for intent resolution
 	isRange                // range commands may span multiple keys
+	isReverse              // reverse commands traverse ranges in descending direction
 )
 
 // IsAdmin returns true if the request requires admin permissions.
 func IsAdmin(args Request) bool {
+	if br, ok := args.(*BatchRequest); ok {
+		return br.IsAdmin()
+	}
 	return (args.flags() & isAdmin) != 0
 }
 
 // IsRead returns true if the request requires read permissions.
 func IsRead(args Request) bool {
+	if br, ok := args.(*BatchRequest); ok {
+		return br.IsRead()
+	}
 	return (args.flags() & isRead) != 0
 }
 
 // IsWrite returns true if the request requires write permissions.
 func IsWrite(args Request) bool {
+	if br, ok := args.(*BatchRequest); ok {
+		return br.IsWrite()
+	}
 	return (args.flags() & isWrite) != 0
 }
 
 // IsReadOnly returns true if the request only requires read
 // permissions.
 func IsReadOnly(args Request) bool {
+	if br, ok := args.(*BatchRequest); ok {
+		return br.IsReadOnly()
+	}
 	return IsRead(args) && !IsWrite(args)
 }
 
+// IsReverse returns true if the request is reverse.
+func IsReverse(args Request) bool {
+	if br, ok := args.(*BatchRequest); ok {
+		return br.IsReverse()
+	}
+	return (args.flags() & isReverse) != 0
+}
+
 // IsTransactionWrite returns true if the request produces write
 // intents when used within a transaction.
 func IsTransactionWrite(args Request) bool {
+	if br, ok := args.(*BatchRequest); ok {
+		return br.IsTransactionWrite()
+	}
 	return (args.flags() & isTxnWrite) != 0
 }
 
 // IsRange returns true if the operation is range-based and must include
 // a start and an end key.
 func IsRange(args Request) bool {
+	if br, ok := args.(*BatchRequest); ok {
+		return br.IsRange()
+	}
 	return (args.flags() & isRange) != 0
 }
 
+// IsAdmin returns true iff the BatchRequest contains an admin request.
+func (br *BatchRequest) IsAdmin() bool {
+	for _, arg := range br.Requests {
+		if IsAdmin(arg.GetValue().(Request)) {
+			return true
+		}
+	}
+	return false
+}
+
+// IsRead returns true if all requests within are flagged as reading data.
+func (br *BatchRequest) IsRead() bool {
+	for i := range br.Requests {
+		if !IsRead(br.Requests[i].GetValue().(Request)) {
+			return false
+		}
+	}
+	return true
+}
+
+// IsWrite returns true iff the BatchRequest contains a write.
+func (br *BatchRequest) IsWrite() bool {
+	for _, arg := range br.Requests {
+		if IsWrite(arg.GetValue().(Request)) {
+			return true
```

Comment:
By which I mean, I'll probably keep `IsWrite`, but it will only apply to the internally chunked batch which can only ever be all writes or all reads, and asserts that there's no mixing.